In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

In [15]:
!pip install imbalanced-learn


  Obtaining dependency information for imbalanced-learn from https://files.pythonhosted.org/packages/92/e8/86c36e1b13007ca9c89381adac6c078cfc8fb71841a76c08a3fe3eca91d3/imbalanced_learn-0.12.0-py3-none-any.whl.metadata
   ---------------------------------------- 0.0/257.7 kB ? eta -:--:--
   - -------------------------------------- 10.2/257.7 kB ? eta -:--:--
   ----------------- ---------------------- 112.6/257.7 kB 1.3 MB/s eta 0:00:01
   ---------------------------------- ----- 225.3/257.7 kB 2.0 MB/s eta 0:00:01
   ---------------------------------------- 257.7/257.7 kB 2.0 MB/s eta 0:00:00


---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
data = pd.read_csv(Path("Resources/lending_data.csv"))

# Review the DataFrame
print(data.head())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = data['loan_status']

# Separate the X variable, the features
X = data.drop('loan_status', axis=1)

In [5]:
# Review the y variable Series
print(y.head())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64


In [6]:
# Review the X variable DataFrame
print(X.head())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  


### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [7]:
# Check the balance of our target values
print(y.value_counts())

loan_status
0    75036
1     2500
Name: count, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using training data
model.fit(X_train, y_train)

C:\Users\johns\.conda\envs\dev\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
predictions = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import balanced_accuracy_score
balanced_accuracy = balanced_accuracy_score(y_test, predictions)
print(f"Balanced Accuracy Score: {balanced_accuracy}")

Balanced Accuracy Score: 0.967989851522121


In [12]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test, predictions)
print(f"Confusion Matrix: \n{matrix}")


Confusion Matrix: 
[[18655   110]
 [   36   583]]


In [13]:
# Print the classification report for the model
from sklearn.metrics import classification_report
report = classification_report(y_test, predictions)
print(f"Classification Report: \n{report}")

Classification Report: 
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.94      0.89       619

    accuracy                           0.99     19384
   macro avg       0.92      0.97      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model seems to predict both the 0 (healthy loan) and 1 (high-risk loan) labels quite well.

For the 0 (healthy loan) label:

The model has a high precision of 1.00, meaning that when it predicts a loan is healthy, it is correct most of the time.
The recall is 0.94, which means it correctly identifies 94% of all actual healthy loans.
For the 1 (high-risk loan) label:

The model has a precision of 0.89, meaning that when it predicts a loan is high-risk, it is correct 89% of the time.
The recall is 0.99, which means it correctly identifies 99% of all actual high-risk loans.
The balanced accuracy score of the model is approximately 0.967, which is quite high. This score is the average of recall obtained on each class, indicating a good balance between sensitivity and specificity.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [16]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

In [17]:
# Count the distinct values of the resampled labels data
print(y_resampled.value_counts())

loan_status
0    56271
1    56271
Name: count, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [18]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
model_resampled.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
predictions_resampled = model_resampled.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_resampled = balanced_accuracy_score(y_test, predictions_resampled)
print(f"Balanced Accuracy Score: {balanced_accuracy_resampled}")

Balanced Accuracy Score: 0.9935981855334257


In [20]:
# Generate a confusion matrix for the model
matrix_resampled = confusion_matrix(y_test, predictions_resampled)
print(f"Confusion Matrix: \n{matrix_resampled}")

Confusion Matrix: 
[[18646   119]
 [    4   615]]


In [21]:
# Print the classification report for the model
report_resampled = classification_report(y_test, predictions_resampled)
print(f"Classification Report: \n{report_resampled}")

Classification Report: 
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model, fit with oversampled data, predicts both the 0 (healthy loan) and 1 (high-risk loan) labels quite well.

For the 0 (healthy loan) label:

The model has a high precision of 1.00, meaning that when it predicts a loan is healthy, it is correct most of the time.
The recall is 0.99, which means it correctly identifies 99% of all actual healthy loans.
For the 1 (high-risk loan) label:

The model has a precision of 0.84, meaning that when it predicts a loan is high-risk, it is correct 84% of the time.
The recall is 0.99, which means it correctly identifies 99% of all actual high-risk loans.
The balanced accuracy score of the model is approximately 0.999, which is quite high. This score is the average of recall obtained on each class, indicating a good balance between sensitivity and specificity.